sklearn/linear_model/_coordinate_descent.py

```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
             "your estimator with the appropriate `alphas=` "
             "argument."
         )
-    n_samples = len(y)
-
-    sparse_center = False
-    if Xy is None:
-        X_sparse = sparse.isspmatrix(X)
-        sparse_center = X_sparse and (fit_intercept or normalize)
-        X = check_array(
-            X, accept_sparse="csc", copy=(copy_X and fit_intercept and not X_sparse)
+    if Xy is not None and ((sample_weight is None) or (fit_intercept is False)):
+        # In this case, the precomputed Xy should be valid.
+        pass
+    else:
+        # Compute Xy.
+        X, y, X_offset, y_offset, X_scale = _preprocess_data(
+            X,
+            y,
+            fit_intercept,
+            normalize=normalize,
+            copy=copy_X,
+            sample_weight=sample_weight,
+            check_input=False,
         )
-        if not X_sparse:
-            # X can be touched inplace thanks to the above line
-            X, y, _, _, _ = _preprocess_data(X, y, fit_intercept, normalize, copy=False)
-        Xy = safe_sparse_dot(X.T, y, dense_output=True)
-
-        if sparse_center:
-            # Workaround to find alpha_max for sparse matrices.
-            # since we should not destroy the sparsity of such matrices.
-            _, _, X_offset, _, X_scale = _preprocess_data(
-                X, y, fit_intercept, normalize
-            )
-            mean_dot = X_offset * np.sum(y)
+        if sample_weight is not None:
+            yw = y * sample_weight / sample_weight.mean()
```

Comment:
```suggestion
            yw = y * sample_weight
```
Otherwise, the unweighted case should read `yw = y / n_samples` instead of `yw = y` because it must be equivalent to all sample weights equal `1`.

Reply:
I chose `sample_weight.mean()` rather than `sample_weight.sum()` for this reason.

Reply:
I'll be honest with you. I'm just dividing by `n_samples` because it's what the old code did, and because it passes the tests.
There are some comments in the `ElasticNet` class that seem to explain it:
```python
# But enet_path coordinate descent minimizes
#
#     1/2 * sum(squared error) + alpha' * penalty                  (2)
#
# and therefore sets
#
#     alpha' = n_samples * alpha                                   (3)
```